{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "IGV Dash Demo",
      "provenance": [],
      "authorship_tag": "ABX9TyNkOmgw36ieOA0M8cbRo9SX",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/broadinstitute/igv-experiments/blob/main/igv-dash/IGV_Dash_Demo.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Interactive IGV Technology Demo\n",
        "\n",
        "## This notebook demonstrates how to use [Dash](https://plotly.com/dash/) to create a plot that will drive the current position within an embedded instance of [IGV](https://igv.org/)."
      ],
      "metadata": {
        "id": "2xDjUcAD0ihc"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "vDG8mSO0aIsm",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "#@markdown # Step One: Build and Configure\n",
        "#@markdown This demo builds its own version of the python package\n",
        "#@markdown [cyvcf2](https://github.com/brentp/cyvcf2), so it can take up to two minutes to complete this \n",
        "#@markdown initialization cell. \n",
        "\n",
        "def install_cyvcf2(path_lib='./lib'):\n",
        "    import sys, subprocess, os\n",
        "    cyvcf_install_script = f\"\"\"\n",
        "    set -e\n",
        "    package=cyvcf2\n",
        "    cwd=$(pwd)\n",
        "    cd /content\n",
        "    rm -rf $package\n",
        "    if [ ! -d $package ]; then\n",
        "    rm -rf $package\n",
        "    fi\n",
        "    git clone --recursive https://github.com/brentp/$package\n",
        "    apt-get install autoconf\n",
        "    cd $package/htslib\n",
        "    autoheader\n",
        "    autoconf\n",
        "    ./configure --enable-gcs --enable-libcurl\n",
        "    make \n",
        "    cd .. \n",
        "    pip install -r requirements.txt \n",
        "    CYTHONIZE=1 pip install -e .\n",
        "    cd ..\n",
        "    cp -r $package {path_lib}\n",
        "    rm -rf $package\n",
        "    cd $cwd\n",
        "    \"\"\"\n",
        "    \n",
        "    cyvcf2_path = f\"{path_lib}/cyvcf2\"\n",
        "    if not os.path.exists(cyvcf2_path):\n",
        "        print(\"Building python cyvcf2\")\n",
        "        build_proc = subprocess.Popen(cyvcf_install_script, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)\n",
        "        with open('cyvcf2-build.log', 'w') as fh:\n",
        "            for line in iter(build_proc.stdout.readline, b''):\n",
        "                line = line.decode().strip()\n",
        "                fh.write(line)\n",
        "    \n",
        "    sys.path.append(cyvcf2_path)\n",
        "    \n",
        "def configure_dash_proxy(port=8050):\n",
        "    import os\n",
        "    server_url = None\n",
        "    base_subpath = None\n",
        "\n",
        "    # Terra Notebooks\n",
        "    gcs_project = os.environ.get(\"GOOGLE_PROJECT\")\n",
        "    runtime_name = os.environ.get(\"RUNTIME_NAME\")\n",
        "    if (gcs_project and runtime_name):\n",
        "        base_subpath = f'/notebooks/{gcs_project}/{runtime_name}/'\n",
        "        server_url = f'https://notebooks.firecloud.org/proxy/{gcs_project}/{runtime_name}/jupyter/proxy/{port}/'\n",
        "    else:\n",
        "        # Google Colab\n",
        "        try:\n",
        "            from google.colab.output import eval_js\n",
        "            js_code = f\"google.colab.kernel.proxyPort({port})\"\n",
        "            server_url = eval_js(js_code)\n",
        "            base_subpath = '/'\n",
        "        except ImportError:\n",
        "            pass\n",
        "\n",
        "    if server_url and base_subpath:\n",
        "        from jupyter_dash.comms import _jupyter_config\n",
        "        from jupyter_dash import JupyterDash\n",
        "\n",
        "        _jupyter_config.update({\n",
        "            'type': 'base_url_response',\n",
        "            'server_url': server_url, \n",
        "            'base_subpath': '/',\n",
        "            'frontend': 'notebook'\n",
        "        })\n",
        "        JupyterDash.default_server_url = server_url\n",
        "\n",
        "def init_dash(port=8050):\n",
        "    from importlib.util import find_spec\n",
        "    if find_spec('jupyter_dash') is None:\n",
        "        !pip install jupyter-dash==0.4.2 dash-bio >/dev/null 2>&1\n",
        "        #!pip install jupyter-dash==0.4.1 dash-bio pyyaml==5.4.1 >/dev/null 2>&1\n",
        "    configure_dash_proxy(port=port)\n",
        "    # squelch confusing log messages from app server\n",
        "    import logging\n",
        "    logging.getLogger('werkzeug').setLevel(logging.ERROR)\n",
        "    print(\"jupyter-dash configured\")\n",
        "    return True\n",
        "    \n",
        "install_cyvcf2()\n",
        "_ = init_dash()"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@markdown # Step Two: Sample VCF Data\n",
        "#@markdown Before we can show an interactive plot, we need data first. This cell will sample \n",
        "#@markdown thousands of sites from a VCF for HG001 from GIAB, and cache the data \n",
        "#@markdown into a pandas data frame.\n",
        "\n",
        "import re\n",
        "import random\n",
        "import pandas as pd\n",
        "from cyvcf2 import VCF\n",
        "\n",
        "class VariantSampler(object):\n",
        "    def random_region(self, vcf, seqmap):\n",
        "        (seqname, seqlen) = random.choice(seqmap)\n",
        "        pos = random.randint(0, seqlen)\n",
        "        coord = f\"{seqname}:{pos}\"\n",
        "        region = vcf(coord)\n",
        "        return region\n",
        "    \n",
        "    def sample(self, vcf_path=None, n_sites=100, n_samples=100, seqmap=None):\n",
        "        vcf_fn = vcf_path.split('/')[-1]\n",
        "        print(f\"Sampling {n_sites * n_samples} variants from {vcf_fn}\")\n",
        "        vcf = VCF(vcf_path)\n",
        "        if seqmap is None:\n",
        "            re_chr = re.compile(\"^chr\\d{1,2}$\")\n",
        "            seqnames = filter(re_chr.match, vcf.seqnames)\n",
        "            seqmap = list(zip(seqnames, vcf.seqlens))\n",
        "\n",
        "        for site_idx in range(n_sites):\n",
        "            sample_cnt = 0\n",
        "            region = self.random_region(vcf, seqmap)\n",
        "            while sample_cnt < n_samples:\n",
        "                try:\n",
        "                    rec = next(region)\n",
        "                except StopIteration:\n",
        "                    region = self.random_region(vcf, seqmap)\n",
        "                    continue\n",
        "                yield rec\n",
        "                sample_cnt += 1\n",
        "\n",
        "def vcf_to_dict(itr):\n",
        "    format_fields = ['DP', 'GQ', 'AD', 'ADALL']\n",
        "    for rec in itr:\n",
        "        row = {\n",
        "            'chr': rec.CHROM,\n",
        "            'pos': rec.POS,\n",
        "            'ref': rec.REF,\n",
        "            'alt': rec.ALT,\n",
        "            'filter': rec.FILTER,\n",
        "        }\n",
        "        info = {key: val.split(',') if (type(val) == str) else val for (key, val) in dict(rec.INFO).items()}\n",
        "        row.update(info)\n",
        "        format_values = {fn: rec.format(fn)[0][0] for fn in format_fields}\n",
        "        row.update(format_values)\n",
        "        yield row\n",
        "\n",
        "def sample_vcf_as_dataframe(vcf_path=None, n_sites=50, n_samples=100):\n",
        "    vs = VariantSampler()\n",
        "    sites = vs.sample(vcf_path, n_sites=n_sites, n_samples=n_samples)\n",
        "    sites = vcf_to_dict(sites)\n",
        "    df = pd.DataFrame(list(sites))\n",
        "    df = df.fillna(0)\n",
        "    #pass_indicies = (df['filter'] == 0)\n",
        "    #df['filter'][pass_indicies] = 'PASS'\n",
        "    df.loc[df['filter'] == 0, 'filter'] = 'PASS'\n",
        "    df['locus'] = [f'{ch}:{pos}' for (ch, pos) in zip(df['chr'], df['pos'])]\n",
        "    df['_index'] = df.index\n",
        "    return df\n",
        "\n",
        "vcf_path = \"https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/NA12878_HG001/latest/GRCh38/SupplementaryFiles/HG001_GRCh38_1_22_v4.2.1_all.vcf.gz\"\n",
        "vcf_index_path = vcf_path + '.tbi'\n",
        "\n",
        "cram_path = \"gs://broad-public-datasets/NA12878/NA12878.cram\"\n",
        "cram_index_path = \"gs://broad-public-datasets/NA12878/NA12878.cram.crai\"\n",
        "\n",
        "vcf_data = sample_vcf_as_dataframe(vcf_path=vcf_path, n_sites=10, n_samples=100)"
      ],
      "metadata": {
        "id": "TuNrgOdhxJHQ",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@markdown # Step Three: Explore Variant Data with Plotly and IGV\n",
        "#@markdown Now that we have our sampled variant data, lets explore it. We will create a scatter plot of all the variants, with the X axis representing variant quality (GQ) and the Y axis representing variant diversity within the larger cohort (ADALL). The color of the marker indicates the variant's filter status. Once the graph loads, click on any marker to focus IGV on the variant in question. On the right side of the scatter plot is a data pane that shows tabular information.\n",
        "#@markdown This cell will output a URL that points to your plotly applicaton.\n",
        "\n",
        "import logging\n",
        "import json\n",
        "import dash\n",
        "import dash_bio\n",
        "import plotly.express as px\n",
        "from jupyter_dash import JupyterDash\n",
        "from dash import dcc, html\n",
        "from dash.dependencies import Input, Output, State\n",
        "from dash.exceptions import PreventUpdate\n",
        "\n",
        "igv_tracks = [\n",
        "    {\n",
        "        \"name\": \"HG001 Variants\",\n",
        "        \"url\": vcf_path,\n",
        "        \"indexURL\": vcf_index_path,\n",
        "        \"format\": \"vcf\",\n",
        "        \"displayMode\": \"SQUISHED\",\n",
        "    },\n",
        "    {\n",
        "        \"name\": \"HG001 Alignments\",\n",
        "        \"url\": cram_path,\n",
        "        \"indexURL\": cram_index_path,\n",
        "        \"format\": \"cram\",\n",
        "        \"displayMode\": \"SQUISHED\",\n",
        "    },\n",
        "    {\n",
        "        \"name\": \"Common SNPs (150)\",\n",
        "        \"type\": \"snp\",\n",
        "        \"format\": \"snp\",\n",
        "        \"url\": \"https://s3.dualstack.us-east-1.amazonaws.com/igv.org.genomes/hg38/annotations/snp150Common.txt.gz\",\n",
        "        \"indexURL\": \"https://s3.dualstack.us-east-1.amazonaws.com/igv.org.genomes/hg38/annotations/snp150Common.txt.gz.tbi\",\n",
        "        \"visibilityWindow\": 100000,\n",
        "    },\n",
        "]\n",
        "\n",
        "\n",
        "# Build App\n",
        "app = JupyterDash(__name__)\n",
        "app.layout = html.Div([\n",
        "    html.H1(\"VCF IGV Scatter Demo\"),\n",
        "    html.H2(\"To interact with the demo, click on a marker within the scatter plot.\"),\n",
        "    html.Table([\n",
        "        html.Tr([\n",
        "            html.Td([\n",
        "                dcc.Loading(type='default', id='graph-loading', children=dcc.Graph(id='graph-output')),\n",
        "            ], style={'width': '1024px'}),\n",
        "            html.Td([\n",
        "                html.Pre(id=\"where\")\n",
        "            ]),\n",
        "        ]),\n",
        "    ]),\n",
        "    dcc.Loading(type='default', id='igv-loading', children=html.Div(id='igv-output')),\n",
        "], style={\"background-color\": \"white\"})\n",
        "\n",
        "# Define callback to update graph\n",
        "@app.callback(\n",
        "    Output('graph-output', 'figure'),\n",
        "    Input('graph-loading', 'children')\n",
        ")\n",
        "def update_figure(children):\n",
        "    return px.scatter(\n",
        "        vcf_data,\n",
        "        x=\"GQ\", \n",
        "        y=\"ADALL\",\n",
        "        color=\"filter\",\n",
        "        size=\"ADALL\",\n",
        "        custom_data=[\"_index\"],\n",
        "        render_mode=\"webgl\", \n",
        "        title=\"HG001: ADALL vs GQ\",\n",
        "    )\n",
        "\n",
        "@app.callback(\n",
        "    Output('igv-output', 'children'),\n",
        "    Input('graph-output', 'clickData')\n",
        ")\n",
        "def return_igv(clickData):\n",
        "    igv_config = dict(\n",
        "        id='igv-chart',\n",
        "        genome='hg38',\n",
        "        tracks=igv_tracks,\n",
        "    )\n",
        "        \n",
        "    if clickData:\n",
        "        idx = clickData['points'][0]['customdata'][0]\n",
        "        locus = vcf_data.loc[idx]['locus']\n",
        "        igv_config[\"locus\"] = locus\n",
        "    \n",
        "    igv = dash_bio.Igv(**igv_config)\n",
        "    dom = html.Div([igv])\n",
        "    return dom\n",
        "    \n",
        "@app.callback(\n",
        "    Output('where', 'children'),\n",
        "    Input('graph-output', 'clickData')\n",
        ")\n",
        "def click_data(clickData):\n",
        "    if not clickData:\n",
        "        return ''\n",
        "    idx = clickData['points'][0]['customdata'][0]\n",
        "    return repr(vcf_data.loc[idx])\n",
        "\n",
        "# Run app and display result inline in the notebook\n",
        "debug = False\n",
        "mode = 'external'\n",
        "_ = app.run_server(debug=debug, mode=mode, dev_tools_ui=debug, dev_tools_serve_dev_bundles=debug)"
      ],
      "metadata": {
        "id": "S5OquyQvFIqH",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}